src/backend/tcop/postgres.c

```diff
@@ -5303,7 +5308,38 @@ PostgresMain(int argc, char *argv[],
 		if (ConfigReloadPending)
 		{
 			ConfigReloadPending = false;
-			ProcessConfigFile(PGC_SIGHUP);
+
+			/*
+			 * On QE backends, GUC options are not actually read from the file,
+			 * but rather sent to us by QD. It means that default settings on
+			 * segments have source of PGC_S_CLIENT instead of PGC_S_FILE, which
+			 * is higher.
+			 *
+			 * We repeat the process and send the GUCs to the QEs again.
+			 * NOTE: This will will increase changed GUCs' source from
+			 * PGC_S_CLIENT all the way up to PGC_S_SESSION.
+			 */
+			if (Gp_role != GP_ROLE_DISPATCH)
+				ProcessConfigFile(PGC_SIGHUP);
+			else
+			{
+				ListCell   *lc;
+				List	   *changed_gucs = NIL;
+				List	   *changed_guc_names = ProcessConfigFileForSync(PGC_SIGHUP);
+
+				foreach (lc, changed_guc_names)
+				{
+					struct config_generic *guc = find_option(lfirst(lc), false, DEBUG2);
+					if (guc != NULL)
+						changed_gucs = lappend(changed_gucs, guc);
+				}
+
+				if (changed_gucs != NIL)
+					send_guc_to_QE(changed_gucs);
+
+				list_free(changed_guc_names);
+				list_free(changed_gucs);
```

Comment:
You can move this list_free and the other one below into the send_guc_to_QE function.

Reply:
Isn't it more clear when the object is allocated and freed in the same context? I think it will be confusing to delegate the cleanup to a function.